target_benchmark/tasks/AbsTask.py

```diff
@@ -223,37 +305,59 @@ def task_run(
         logger.info(f"start task {self.task_name}")
 
         for dataset_name, dataset_loader in dataset_loaders.items():
+            # construct the path to persistence files
+            path_to_retrieval_results = construct_persistence_path(path_to_retrieval_results_dir, dataset_name, top_k)
+            path_to_downstream_results = construct_persistence_path(path_to_downstream_results_dir, dataset_name, top_k)
+
+            # construct generators
+            prev_retrieval_res_gen = generate_batches_from_file(
+                path_to_retrieval_results,
+                batch_size,
+                RetrievalResultDataModel,
+            )
+            prev_downstream_res_gen = generate_batches_from_file(
+                path_to_downstream_results,
+                batch_size,
+                DownstreamGeneratedResultDataModel,
+            )
+
             logger.info(f"running task on dataset {dataset_name}")
+
             table_id_to_table = dataset_loader.get_table_id_to_table()
+            # some retrieval metrics to track
             total_process_duration = 0
             total_wall_clock_duration = 0
+            total_num_retrieved = 0
+
+            # set up progress bar
             total_num_queries = dataset_loader.get_queries_size()
             progress_bar = tqdm(total=total_num_queries, desc=f"Retrieving Tables for {dataset_name}...")
-            for query_batch in dataset_loader.get_queries_for_task(batch_size):
-                (
-                    retrieval_results,
-                    process_duration,
-                    wall_clock_duration,
-                ) = self._get_retrieval_results(
+            for query_batch in dataset_loader.get_queries_for_task(batch_size=batch_size):
+                # run retrieval on batch
+                retrieval_results, process_duration, wall_clock_duration, num_retrieved = self._run_retrieval_batch(
                     retriever,
-                    query_batch,
                     dataset_name,
+                    query_batch,
                     top_k,
+                    prev_retrieval_res_gen,
+                    path_to_retrieval_results,
                     **kwargs,
                 )
+
+                # update time spent
                 total_process_duration += process_duration
                 total_wall_clock_duration += wall_clock_duration
-                self._update_retrieval_metrics(query_batch, retrieval_results)
-                if path_to_retrieval_results:
-                    self._write_results(retrieval_results, path_to_retrieval_results)
+                total_num_retrieved += num_retrieved
 
-                downstream_results = self._get_downstream_task_results(
-                    query_batch, retrieval_results, dataset_name, table_id_to_table
+                # run downstream on batch
+                self._run_downstream_batch(
```

Comment:
General comment: for functions with large signatures, it will save a lot of headache debugging in the future by passing named arguments vs. the positional arguments being done here.  

So, `retrieval_results=retrieval_results`, etc.